library/rsa.c

```diff
@@ -525,40 +531,51 @@ int mbedtls_rsa_gen_key( mbedtls_rsa_context *ctx,
         MBEDTLS_MPI_CHK( mbedtls_mpi_gen_prime( &ctx->Q, nbits >> 1, 0,
                                                 f_rng, p_rng ) );
 
-        if( mbedtls_mpi_cmp_mpi( &ctx->P, &ctx->Q ) == 0 )
-            continue;
-
-        MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &ctx->N, &ctx->P, &ctx->Q ) );
-        if( mbedtls_mpi_bitlen( &ctx->N ) != nbits )
+        /* make sure the difference between p and q is not too small (FIPS 186-4 §B.3.3 step 5.4) */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mpi( &H, &ctx->P, &ctx->Q ) );
+        if( ( nbits >= 200 ) && ( mbedtls_mpi_bitlen( &H ) <= ( nbits >> 1 ) - 100 + 1 ) )
```

Comment:
We must guard against `P = Q` even if `nbits < 200`, which is currently not the case.

Reply:
Done. I'm already doing a comparison on line 535, but I'm not sure how to leverage the result of that again on line 537, so I'm using `mbedtls_mpi_cmp_int`.

Reply:
An alternative way would be to test `mbedtls_mpi_bitlen( &H ) <= ( nbits >= 200 ) ? ( nbits >> 1 ) - 99 : 0`.

Reply:
Done